hypothesis-python/src/hypothesis/internal/conjecture/dfa/__init__.py

```diff
@@ -61,55 +84,183 @@ def matches(self, s):
             i = self.transition(i, c)
         return self.is_accepting(i)
 
+    @cached
+    def max_length(self, i):
+        """Returns the maximum length of a string that is
+        accepted when starting from i."""
+        if self.is_dead(i):
+            return 0
+        if i in self.reachable(i):
+            return inf
+        next_states = {self.max_length(j) for _, j in self.transitions(i)}
+        if next_states:
+            return 1 + max(next_states)
+        else:
+            assert self.is_accepting(i)
+            return 0
+
+    @cached
+    def count_strings(self, i, k):
+        """Returns the number of strings of length ``k``
+        that are accepted when starting from state ``i``."""
+        assert k >= 0
+        if k == 0:
+            if self.is_accepting(i):
+                return 1
+            else:
+                return 0
+        if k > self.max_length(i):
+            return 0
+        return sum(self.count_strings(j, k - 1) for _, j in self.transitions(i))
```

Comment:
One thing I've noticed is that we (currently?) don't actually care about the specific value of `count_strings`; we just care if it's zero or non-zero.

It would therefore be possible to replace this with an almost-identical `has_strings` method that can short-circuit as soon as it sees that one of its descendants has a non-zero number of strings of the appropriate length (i.e. by replacing `sum` with `any`).

(No idea if this would be an actual improvement in practice; I'm fine with leaving things as-is.)

Reply:
Yeah, I'd noticed that and thought about it, but some of the upcoming work needs the full count, so I decided to skip straigh to that. Maintaining just the `has_strings` calculation might be better for space efficiency reasons? Not sure. Probably doesn't matter that much.

Reply:
In that case I'm happy with just having `count_strings` as-is.